pkg/driver/controller.go

```diff
@@ -233,23 +252,60 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 		gid = allocatedGid
 	}
 
+	if value, ok := volumeParams[BasePath]; ok {
+		basePath = value
+	}
+
 	rootDirName := volName
-	rootDir := basePath + "/" + rootDirName
+	// Check if a custom structure should be imposed on the access point directory
+	if value, ok := volumeParams[SubPathPattern]; ok {
+		// Try and construct the root directory and check it only contains supported components
+		val, err := interpolateRootDirectoryName(value, volumeParams)
+		if err == nil {
+			klog.Infof("Using user-specified structure for access point directory.")
+			rootDirName = val
+			if value, ok := volumeParams[EnsureUniqueDirectory]; ok {
+				if ensureUniqueDirectory, err := strconv.ParseBool(value); !ensureUniqueDirectory && err == nil {
+					klog.Infof("Not appending PVC UID to path.")
+				} else {
+					klog.Infof("Appending PVC UID to path.")
+					rootDirName = fmt.Sprintf("%s-%s", val, uuid.New().String())
+				}
+			} else {
+				klog.Infof("Appending PVC UID to path.")
+				rootDirName = fmt.Sprintf("%s-%s", val, uuid.New().String())
+			}
+		} else {
+			return nil, err
+		}
```

Comment:
Early return in case of error.

```suggestion
		if err != nil {
			klog.Infof("Using user-specified structure for access point directory.")			
			return nil, err		
		} 
		rootDirName = val
		if value, ok := volumeParams[EnsureUniqueDirectory]; ok {
			if ensureUniqueDirectory, err := strconv.ParseBool(value); !ensureUniqueDirectory && err == nil {
				klog.Infof("Not appending PVC UID to path.")
			} else {
				klog.Infof("Appending PVC UID to path.")
				rootDirName = fmt.Sprintf("%s-%s", val, uuid.New().String())
			}
		} else {
			klog.Infof("Appending PVC UID to path.")
			rootDirName = fmt.Sprintf("%s-%s", val, uuid.New().String())
		}		
```